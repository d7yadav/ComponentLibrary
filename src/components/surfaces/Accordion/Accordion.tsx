import { ExpandMore, ChevronRight } from '@mui/icons-material';
import type { SyntheticEvent } from 'react';
import React, { forwardRef, memo, useCallback, useMemo } from 'react';

import {
  ACCORDION_VARIANTS,
  ACCORDION_SIZES,
  ACCORDION_ICON_POSITIONS,
  ACCORDION_TRANSITIONS,
  ACCORDION_FOCUS_COLORS,
  ANIMATION_DURATIONS,
  DEFAULT_PROPS,
  ACCESSIBILITY_CONSTANTS,
  TEST_IDS,
} from './Accordion.constants';
import {
  StyledAccordion,
  StyledAccordionSummary,
  StyledAccordionDetails,
  StyledAccordionActions,
  SummaryContent,
  SummaryTitle,
  SummarySubtitle,
  IconWrapper,
  StyledDivider,
  loadingSpinKeyframes,
} from './Accordion.styles';
import type { AccordionProps, AccordionSummaryProps, AccordionDetailsProps, AccordionActionsProps } from './Accordion.types';

/**
 * Enhanced Accordion component with multiple variants, animations, and accessibility features
 * 
 * Features:
 * - 5 variants: standard, outlined, elevated, flat, card
 * - 3 sizes: compact, comfortable, spacious
 * - Flexible icon positioning: start, end, both, none
 * - Smooth animations with customizable timing
 * - Single and multiple expand modes
 * - Comprehensive accessibility (WCAG 2.1 AA)
 * - Custom expand/collapse icons
 * - Summary subtitles and actions
 * - Mobile-optimized touch interactions
 * - Dark theme support
 */
const AccordionComponent = forwardRef<HTMLDivElement, AccordionProps>(({
  variant = DEFAULT_PROPS.variant,
  size = DEFAULT_PROPS.size,
  expandMode = DEFAULT_PROPS.expandMode,
  iconPosition = DEFAULT_PROPS.iconPosition,
  transition = DEFAULT_PROPS.transition,
  expanded = DEFAULT_PROPS.expanded,
  onChange,
  disabled = DEFAULT_PROPS.disabled,
  disableGutters = DEFAULT_PROPS.disableGutters,
  showDivider = DEFAULT_PROPS.showDivider,
  expandIcon,
  collapseIcon,
  startIcon,
  endIcon,
  summary,
  children,
  actions,
  subtitle,
  disableToggle = DEFAULT_PROPS.disableToggle,
  elevation = DEFAULT_PROPS.elevation,
  square = DEFAULT_PROPS.square,
  className,
  summaryClassName,
  detailsClassName,
  actionsClassName,
  style,
  summaryStyle,
  detailsStyle,
  actionsStyle,
  animationDuration = DEFAULT_PROPS.animationDuration,
  dense = DEFAULT_PROPS.dense,
  focusColor = DEFAULT_PROPS.focusColor,
  'data-testid': dataTestId = TEST_IDS.accordion,
  'aria-label': ariaLabel,
  'aria-labelledby': ariaLabelledBy,
  'aria-describedby': ariaDescribedBy,
  ...other
}, ref) => {
  // Handle accordion expansion
  const handleChange = useCallback((event: SyntheticEvent, isExpanded: boolean) => {
    if (disabled || disableToggle) return;
    onChange?.(event, isExpanded);
  }, [onChange, disabled, disableToggle]);

  // Determine expand icon
  const getExpandIcon = useMemo(() => {
    if (iconPosition === ACCORDION_ICON_POSITIONS.none) return null;
    
    if (expandIcon && collapseIcon) {
      return expanded ? collapseIcon : expandIcon;
    }
    
    if (expandIcon) return expandIcon;
    
    // Default icons based on position
    if (iconPosition === ACCORDION_ICON_POSITIONS.start) {
      return <ChevronRight />;
    }
    
    return <ExpandMore />;
  }, [iconPosition, expandIcon, collapseIcon, expanded]);

  // Generate unique IDs for accessibility
  const summaryId = useMemo(() => `accordion-summary-${Math.random().toString(36).substr(2, 9)}`, []);
  const detailsId = useMemo(() => `accordion-details-${Math.random().toString(36).substr(2, 9)}`, []);

  // Accessibility props
  const accordionAccessibilityProps = {
    'aria-label': ariaLabel,
    'aria-labelledby': ariaLabelledBy || summaryId,
    'aria-describedby': ariaDescribedBy || detailsId,
    role: ACCESSIBILITY_CONSTANTS.accordionRole,
  };

  const summaryAccessibilityProps = {
    id: summaryId,
    'aria-controls': detailsId,
    'aria-expanded': expanded,
    'aria-label': typeof summary === 'string' ? summary : ariaLabel || ACCESSIBILITY_CONSTANTS.expandLabel,
  };

  const detailsAccessibilityProps = {
    id: detailsId,
    'aria-labelledby': summaryId,
    role: 'region',
  };

  // Style props for styled components
  const accordionStyleProps = {
    variant,
    size,
    expanded: Boolean(expanded),
    disabled: Boolean(disabled),
    elevation,
    square: Boolean(square),
    transition,
    animationDuration,
  };

  const summaryStyleProps = {
    variant,
    size,
    iconPosition,
    dense: Boolean(dense),
    focusColor,
    showDivider: Boolean(showDivider),
    disabled: Boolean(disabled),
    expanded: Boolean(expanded),
  };

  const detailsStyleProps = {
    variant,
    size,
    expanded: Boolean(expanded),
  };

  const actionsStyleProps = {
    variant,
    size,
    alignLeft: false, // Default to right alignment
  };

  return (
    <>
      <style>{loadingSpinKeyframes}</style>
      <StyledAccordion
        ref={ref}
        expanded={expanded}
        onChange={handleChange}
        disabled={disabled}
        disableGutters={disableGutters}
        square={square}
        elevation={variant === ACCORDION_VARIANTS.elevated ? elevation : 0}
        className={className}
        style={style}
        data-testid={dataTestId}
        {...accordionAccessibilityProps}
        {...other}
      >
        <StyledAccordionSummary
          expandIcon={getExpandIcon}
          className={summaryClassName}
          style={summaryStyle}
          data-testid={`${dataTestId}-${TEST_IDS.summary}`}
          {...summaryStyleProps}
          {...summaryAccessibilityProps}
        >
          {/* Start icon */}
          {startIcon && iconPosition !== ACCORDION_ICON_POSITIONS.none && (
            <IconWrapper 
              position="start" 
              size={size}
              data-testid={`${dataTestId}-${TEST_IDS.startIcon}`}
            >
              {startIcon}
            </IconWrapper>
          )}

          {/* Summary content */}
          <SummaryContent 
            hasSubtitle={Boolean(subtitle)}
            iconPosition={iconPosition}
          >
            <SummaryTitle>
              {summary}
            </SummaryTitle>
            {subtitle && (
              <SummarySubtitle data-testid={`${dataTestId}-${TEST_IDS.subtitle}`}>
                {subtitle}
              </SummarySubtitle>
            )}
          </SummaryContent>

          {/* End icon */}
          {endIcon && iconPosition !== ACCORDION_ICON_POSITIONS.none && (
            <IconWrapper 
              position="end" 
              size={size}
              data-testid={`${dataTestId}-${TEST_IDS.endIcon}`}
            >
              {endIcon}
            </IconWrapper>
          )}
        </StyledAccordionSummary>

        {/* Divider between summary and details */}
        {showDivider && expanded && (
          <StyledDivider data-testid={`${dataTestId}-${TEST_IDS.divider}`} />
        )}

        <StyledAccordionDetails
          className={detailsClassName}
          style={detailsStyle}
          data-testid={`${dataTestId}-${TEST_IDS.details}`}
          {...detailsStyleProps}
          {...detailsAccessibilityProps}
        >
          {children}
        </StyledAccordionDetails>

        {/* Actions */}
        {actions && (
          <StyledAccordionActions
            className={actionsClassName}
            style={actionsStyle}
            data-testid={`${dataTestId}-${TEST_IDS.actions}`}
            {...actionsStyleProps}
          >
            {actions}
          </StyledAccordionActions>
        )}
      </StyledAccordion>
    </>
  );
});

AccordionComponent.displayName = 'Accordion';

// Export memoized component for performance optimization
export const Accordion = memo(AccordionComponent);

// Enhanced AccordionSummary component for standalone use
const AccordionSummaryComponent = forwardRef<HTMLDivElement, AccordionSummaryProps>(({
  variant = DEFAULT_PROPS.variant,
  size = DEFAULT_PROPS.size,
  iconPosition = DEFAULT_PROPS.iconPosition,
  expandIcon,
  startIcon,
  endIcon,
  dense = DEFAULT_PROPS.dense,
  focusColor = DEFAULT_PROPS.focusColor,
  showDivider = DEFAULT_PROPS.showDivider,
  subtitle,
  children,
  ...other
}, ref) => {
  const styleProps = {
    variant,
    size,
    iconPosition,
    dense,
    focusColor,
    showDivider,
    disabled: false,
    expanded: false,
  };

  return (
    <StyledAccordionSummary
      ref={ref}
      expandIcon={expandIcon}
      {...styleProps}
      {...other}
    >
      {startIcon && (
        <IconWrapper position="start" size={size}>
          {startIcon}
        </IconWrapper>
      )}
      
      <SummaryContent hasSubtitle={Boolean(subtitle)} iconPosition={iconPosition}>
        <SummaryTitle>{children}</SummaryTitle>
        {subtitle && <SummarySubtitle>{subtitle}</SummarySubtitle>}
      </SummaryContent>
      
      {endIcon && (
        <IconWrapper position="end" size={size}>
          {endIcon}
        </IconWrapper>
      )}
    </StyledAccordionSummary>
  );
});

AccordionSummaryComponent.displayName = 'AccordionSummary';

export const AccordionSummary = memo(AccordionSummaryComponent);

// Enhanced AccordionDetails component for standalone use
const AccordionDetailsComponent = forwardRef<HTMLDivElement, AccordionDetailsProps>(({
  variant = DEFAULT_PROPS.variant,
  size = DEFAULT_PROPS.size,
  children,
  ...other
}, ref) => {
  const styleProps = {
    variant,
    size,
    expanded: true,
  };

  return (
    <StyledAccordionDetails
      ref={ref}
      {...styleProps}
      {...other}
    >
      {children}
    </StyledAccordionDetails>
  );
});

AccordionDetailsComponent.displayName = 'AccordionDetails';

export const AccordionDetails = memo(AccordionDetailsComponent);

// Enhanced AccordionActions component for standalone use
const AccordionActionsComponent = forwardRef<HTMLDivElement, AccordionActionsProps>(({
  variant = DEFAULT_PROPS.variant,
  size = DEFAULT_PROPS.size,
  alignLeft = false,
  children,
  ...other
}, ref) => {
  const styleProps = {
    variant,
    size,
    alignLeft,
  };

  return (
    <StyledAccordionActions
      ref={ref}
      {...styleProps}
      {...other}
    >
      {children}
    </StyledAccordionActions>
  );
});

AccordionActionsComponent.displayName = 'AccordionActions';

export const AccordionActions = memo(AccordionActionsComponent);