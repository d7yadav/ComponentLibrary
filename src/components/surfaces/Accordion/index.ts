export { 
  Accordion, 
  AccordionSummary, 
  AccordionDetails, 
  AccordionActions 
} from './Accordion';

export type {
  AccordionProps,
  AccordionSummaryProps,
  AccordionDetailsProps,
  AccordionActionsProps,
  AccordionGroupProps,
  AccordionVariant,
  AccordionSize,
  AccordionExpandMode,
  AccordionIconPosition,
  AccordionTransition,
  AccordionStyleProps,
  AccordionSummaryStyleProps,
  AccordionDetailsStyleProps,
  AccordionActionsStyleProps,
} from './Accordion.types';

export {
  ACCORDION_VARIANTS,
  ACCORDION_SIZES,
  ACCORDION_EXPAND_MODES,
  ACCORDION_ICON_POSITIONS,
  ACCORDION_TRANSITIONS,
  ACCORDION_FOCUS_COLORS,
  ANIMATION_DURATIONS,
  SIZE_CONFIGS,
  ELEVATION_VALUES,
  BORDER_RADIUS,
  TRANSITION_EASINGS,
  Z_INDEX,
  ACCESSIBILITY_CONSTANTS,
  DEFAULT_PROPS,
  TEST_IDS,
  CSS_VARIABLES,
  PERFORMANCE_CONFIG,
} from './Accordion.constants';