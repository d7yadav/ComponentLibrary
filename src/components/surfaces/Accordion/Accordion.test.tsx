/**
 * @fileoverview Accordion Component Tests
 * @author AI Generated Tests - 2025-07-13
 */

import { ThemeProvider , createTheme } from '@mui/material/styles';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { axe, toHaveNoViolations } from 'jest-axe';

import { Accordion, AccordionSummary, AccordionDetails, AccordionActions } from './Accordion';
import type { AccordionProps } from './Accordion.types';

// Add jest-axe matcher
expect.extend(toHaveNoViolations);

// Test theme
const testTheme = createTheme();

// Test wrapper component
const TestWrapper = ({ children }: { children: React.ReactNode }) => (
  <ThemeProvider theme={testTheme}>{children}</ThemeProvider>
);

// Helper function to render component with theme
const renderWithTheme = (ui: React.ReactElement, options = {}) => {
  return render(ui, { wrapper: TestWrapper, ...options });
};

describe('Accordion Component', () => {
  // Basic rendering tests
  describe('Rendering', () => {
    it('renders without crashing', () => {
      renderWithTheme(
        <Accordion>
          <AccordionSummary>Summary</AccordionSummary>
          <AccordionDetails>Details</AccordionDetails>
        </Accordion>
      );
      expect(screen.getByText('Summary')).toBeInTheDocument();
      expect(screen.getByText('Details')).toBeInTheDocument();
    });

    it('renders children correctly', () => {
      renderWithTheme(
        <Accordion>
          <AccordionSummary>Test Summary</AccordionSummary>
          <AccordionDetails>Test Details Content</AccordionDetails>
        </Accordion>
      );
      expect(screen.getByText('Test Summary')).toBeInTheDocument();
      expect(screen.getByText('Test Details Content')).toBeInTheDocument();
    });

    it('renders with custom test id', () => {
      const testId = 'custom-accordion';
      renderWithTheme(
        <Accordion data-testid={testId}>
          <AccordionSummary>Summary</AccordionSummary>
          <AccordionDetails>Details</AccordionDetails>
        </Accordion>
      );
      expect(screen.getByTestId(testId)).toBeInTheDocument();
    });

    it('applies custom className', () => {
      const customClass = 'custom-class';
      renderWithTheme(
        <Accordion className={customClass} data-testid="accordion">
          <AccordionSummary>Summary</AccordionSummary>
          <AccordionDetails>Details</AccordionDetails>
        </Accordion>
      );
      expect(screen.getByTestId('accordion')).toHaveClass(customClass);
    });
  });

  // Props testing
  describe('Props', () => {
    describe('variant prop', () => {
      const variants: AccordionProps['variant'][] = [
        'standard', 'outlined', 'filled', 'elevated', 'ghost'
      ];

      variants.forEach((variant) => {
        it(`renders ${variant} variant correctly`, () => {
          renderWithTheme(
            <Accordion variant={variant} data-testid="variant-test">
              <AccordionSummary>{variant} variant</AccordionSummary>
              <AccordionDetails>Details</AccordionDetails>
            </Accordion>
          );
          const element = screen.getByTestId('variant-test');
          expect(element).toBeInTheDocument();
        });
      });
    });

    describe('size prop', () => {
      const sizes: AccordionProps['size'][] = [
        'small', 'medium', 'large'
      ];

      sizes.forEach((size) => {
        it(`renders ${size} size correctly`, () => {
          renderWithTheme(
            <Accordion size={size} data-testid="size-test">
              <AccordionSummary>{size} size</AccordionSummary>
              <AccordionDetails>Details</AccordionDetails>
            </Accordion>
          );
          const element = screen.getByTestId('size-test');
          expect(element).toBeInTheDocument();
        });
      });
    });

    it('handles expanded state correctly', () => {
      renderWithTheme(
        <Accordion expanded data-testid="expanded-test">
          <AccordionSummary>Expanded Summary</AccordionSummary>
          <AccordionDetails>Expanded Details</AccordionDetails>
        </Accordion>
      );
      
      const button = screen.getByRole('button');
      expect(button).toHaveAttribute('aria-expanded', 'true');
    });

    it('handles disabled state correctly', () => {
      renderWithTheme(
        <Accordion disabled data-testid="disabled-test">
          <AccordionSummary>Disabled Summary</AccordionSummary>
          <AccordionDetails>Disabled Details</AccordionDetails>
        </Accordion>
      );
      
      const button = screen.getByRole('button');
      expect(button).toBeDisabled();
    });

    it('handles controlled mode correctly', () => {
      const handleChange = vi.fn();
      renderWithTheme(
        <Accordion expanded={false} onChange={handleChange}>
          <AccordionSummary>Controlled Summary</AccordionSummary>
          <AccordionDetails>Controlled Details</AccordionDetails>
        </Accordion>
      );
      
      const button = screen.getByRole('button');
      expect(button).toHaveAttribute('aria-expanded', 'false');
    });

    it('handles square prop correctly', () => {
      renderWithTheme(
        <Accordion square data-testid="square-test">
          <AccordionSummary>Square Summary</AccordionSummary>
          <AccordionDetails>Square Details</AccordionDetails>
        </Accordion>
      );
      const element = screen.getByTestId('square-test');
      expect(element).toBeInTheDocument();
    });

    it('handles disableGutters prop correctly', () => {
      renderWithTheme(
        <Accordion disableGutters data-testid="no-gutters-test">
          <AccordionSummary>No Gutters Summary</AccordionSummary>
          <AccordionDetails>No Gutters Details</AccordionDetails>
        </Accordion>
      );
      const element = screen.getByTestId('no-gutters-test');
      expect(element).toBeInTheDocument();
    });
  });

  // Interaction testing
  describe('Interactions', () => {
    it('expands and collapses on click', async () => {
      const user = userEvent.setup();
      
      renderWithTheme(
        <Accordion>
          <AccordionSummary>Clickable Summary</AccordionSummary>
          <AccordionDetails>Collapsible Details</AccordionDetails>
        </Accordion>
      );
      
      const button = screen.getByRole('button');
      
      // Initially collapsed
      expect(button).toHaveAttribute('aria-expanded', 'false');
      
      // Expand
      await user.click(button);
      expect(button).toHaveAttribute('aria-expanded', 'true');
      
      // Collapse
      await user.click(button);
      expect(button).toHaveAttribute('aria-expanded', 'false');
    });

    it('handles keyboard navigation correctly', async () => {
      const user = userEvent.setup();
      
      renderWithTheme(
        <Accordion>
          <AccordionSummary>Keyboard Summary</AccordionSummary>
          <AccordionDetails>Keyboard Details</AccordionDetails>
        </Accordion>
      );
      
      const button = screen.getByRole('button');
      
      // Focus the button
      button.focus();
      expect(button).toHaveFocus();
      
      // Expand with Enter key
      await user.keyboard('{Enter}');
      expect(button).toHaveAttribute('aria-expanded', 'true');
      
      // Collapse with Space key
      await user.keyboard(' ');
      expect(button).toHaveAttribute('aria-expanded', 'false');
    });

    it('calls onChange when expanded state changes', async () => {
      const user = userEvent.setup();
      const handleChange = vi.fn();
      
      renderWithTheme(
        <Accordion onChange={handleChange}>
          <AccordionSummary>Summary</AccordionSummary>
          <AccordionDetails>Details</AccordionDetails>
        </Accordion>
      );
      
      const button = screen.getByRole('button');
      await user.click(button);
      
      expect(handleChange).toHaveBeenCalledTimes(1);
      expect(handleChange).toHaveBeenCalledWith(expect.any(Object), true);
    });

    it('does not respond to interactions when disabled', async () => {
      const user = userEvent.setup();
      const handleChange = vi.fn();
      
      renderWithTheme(
        <Accordion disabled onChange={handleChange}>
          <AccordionSummary>Disabled Summary</AccordionSummary>
          <AccordionDetails>Disabled Details</AccordionDetails>
        </Accordion>
      );
      
      const button = screen.getByRole('button');
      await user.click(button);
      
      expect(handleChange).not.toHaveBeenCalled();
      expect(button).toHaveAttribute('aria-expanded', 'false');
    });

    it('handles hover states correctly', () => {
      renderWithTheme(
        <Accordion>
          <AccordionSummary>Hoverable Summary</AccordionSummary>
          <AccordionDetails>Details</AccordionDetails>
        </Accordion>
      );
      
      const button = screen.getByRole('button');
      fireEvent.mouseEnter(button);
      expect(button).toBeInTheDocument();
      
      fireEvent.mouseLeave(button);
      expect(button).toBeInTheDocument();
    });
  });

  // Subcomponent testing
  describe('Subcomponents', () => {
    describe('AccordionSummary', () => {
      it('renders with expand icon', () => {
        renderWithTheme(
          <Accordion>
            <AccordionSummary expandIcon={<span data-testid="custom-icon">+</span>}>
              Summary with custom icon
            </AccordionSummary>
            <AccordionDetails>Details</AccordionDetails>
          </Accordion>
        );
        
        expect(screen.getByTestId('custom-icon')).toBeInTheDocument();
      });

      it('handles icon position prop', () => {
        renderWithTheme(
          <Accordion>
            <AccordionSummary iconPosition="start" data-testid="icon-start">
              Summary with icon at start
            </AccordionSummary>
            <AccordionDetails>Details</AccordionDetails>
          </Accordion>
        );
        
        expect(screen.getByTestId('icon-start')).toBeInTheDocument();
      });
    });

    describe('AccordionDetails', () => {
      it('renders details content correctly', () => {
        renderWithTheme(
          <Accordion expanded>
            <AccordionSummary>Summary</AccordionSummary>
            <AccordionDetails>
              <div>
                <h3>Details Title</h3>
                <p>Details paragraph</p>
              </div>
            </AccordionDetails>
          </Accordion>
        );
        
        expect(screen.getByText('Details Title')).toBeInTheDocument();
        expect(screen.getByText('Details paragraph')).toBeInTheDocument();
      });
    });

    describe('AccordionActions', () => {
      it('renders action buttons correctly', async () => {
        const user = userEvent.setup();
        const handleSave = vi.fn();
        const handleCancel = vi.fn();
        
        renderWithTheme(
          <Accordion expanded>
            <AccordionSummary>Summary</AccordionSummary>
            <AccordionDetails>Details</AccordionDetails>
            <AccordionActions>
              <button onClick={handleCancel}>Cancel</button>
              <button onClick={handleSave}>Save</button>
            </AccordionActions>
          </Accordion>
        );
        
        const saveButton = screen.getByText('Save');
        const cancelButton = screen.getByText('Cancel');
        
        await user.click(saveButton);
        expect(handleSave).toHaveBeenCalledTimes(1);
        
        await user.click(cancelButton);
        expect(handleCancel).toHaveBeenCalledTimes(1);
      });
    });
  });

  // Accessibility testing
  describe('Accessibility', () => {
    it('meets WCAG accessibility guidelines', async () => {
      const { container } = renderWithTheme(
        <Accordion>
          <AccordionSummary>Accessible Summary</AccordionSummary>
          <AccordionDetails>Accessible Details</AccordionDetails>
        </Accordion>
      );
      
      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });

    it('provides proper ARIA attributes', () => {
      renderWithTheme(
        <Accordion>
          <AccordionSummary>Summary</AccordionSummary>
          <AccordionDetails>Details</AccordionDetails>
        </Accordion>
      );
      
      const button = screen.getByRole('button');
      expect(button).toHaveAttribute('aria-expanded');
      expect(button).toHaveAttribute('aria-controls');
    });

    it('manages focus correctly', async () => {
      const user = userEvent.setup();
      
      renderWithTheme(
        <Accordion>
          <AccordionSummary>Summary</AccordionSummary>
          <AccordionDetails>
            <button>Focusable content</button>
          </AccordionDetails>
        </Accordion>
      );
      
      const summaryButton = screen.getByRole('button', { name: /summary/i });
      const detailsButton = screen.getByRole('button', { name: /focusable content/i });
      
      // Expand accordion
      await user.click(summaryButton);
      
      // Focus should be manageable within details
      detailsButton.focus();
      expect(detailsButton).toHaveFocus();
    });

    it('supports screen readers with proper semantics', () => {
      renderWithTheme(
        <Accordion>
          <AccordionSummary>Accordion Summary</AccordionSummary>
          <AccordionDetails>Accordion Details</AccordionDetails>
        </Accordion>
      );
      
      const button = screen.getByRole('button');
      expect(button).toBeInTheDocument();
      
      // Details should be associated with summary
      const detailsId = button.getAttribute('aria-controls');
      expect(detailsId).toBeTruthy();
    });
  });

  // Theme integration testing
  describe('Theme Integration', () => {
    it('respects theme customizations', () => {
      const customTheme = createTheme({
        components: {
          MuiAccordion: {
            styleOverrides: {
              root: {
                backgroundColor: 'red',
              },
            },
          },
        },
      });

      render(
        <ThemeProvider theme={customTheme}>
          <Accordion data-testid="themed-accordion">
            <AccordionSummary>Themed Summary</AccordionSummary>
            <AccordionDetails>Themed Details</AccordionDetails>
          </Accordion>
        </ThemeProvider>
      );

      expect(screen.getByTestId('themed-accordion')).toBeInTheDocument();
    });

    it('supports dark mode', () => {
      const darkTheme = createTheme({
        palette: {
          mode: 'dark',
        },
      });

      render(
        <ThemeProvider theme={darkTheme}>
          <Accordion data-testid="dark-accordion">
            <AccordionSummary>Dark Summary</AccordionSummary>
            <AccordionDetails>Dark Details</AccordionDetails>
          </Accordion>
        </ThemeProvider>
      );

      expect(screen.getByTestId('dark-accordion')).toBeInTheDocument();
    });
  });

  // Performance testing
  describe('Performance', () => {
    it('renders efficiently with many accordions', () => {
      const manyAccordions = Array.from({ length: 20 }, (_, i) => (
        <Accordion key={i}>
          <AccordionSummary>Summary {i}</AccordionSummary>
          <AccordionDetails>Details {i}</AccordionDetails>
        </Accordion>
      ));

      const { container } = renderWithTheme(<div>{manyAccordions}</div>);
      expect(container).toBeInTheDocument();
    });

    it('handles expansion animations efficiently', async () => {
      const user = userEvent.setup();
      
      renderWithTheme(
        <Accordion>
          <AccordionSummary>Summary</AccordionSummary>
          <AccordionDetails>
            <div style={{ height: '1000px' }}>Large content</div>
          </AccordionDetails>
        </Accordion>
      );
      
      const button = screen.getByRole('button');
      
      // Rapid expansion/collapse should work smoothly
      await user.click(button);
      await user.click(button);
      await user.click(button);
      
      expect(button).toBeInTheDocument();
    });
  });

  // Edge cases and error handling
  describe('Edge Cases', () => {
    it('handles empty accordion gracefully', () => {
      renderWithTheme(<Accordion data-testid="empty-accordion" />);
      expect(screen.getByTestId('empty-accordion')).toBeInTheDocument();
    });

    it('handles accordion with only summary', () => {
      renderWithTheme(
        <Accordion>
          <AccordionSummary>Only Summary</AccordionSummary>
        </Accordion>
      );
      expect(screen.getByText('Only Summary')).toBeInTheDocument();
    });

    it('handles accordion with only details', () => {
      renderWithTheme(
        <Accordion>
          <AccordionDetails>Only Details</AccordionDetails>
        </Accordion>
      );
      expect(screen.getByText('Only Details')).toBeInTheDocument();
    });

    it('handles complex nested content', () => {
      renderWithTheme(
        <Accordion>
          <AccordionSummary>Complex Summary</AccordionSummary>
          <AccordionDetails>
            <div>
              <form>
                <input type="text" placeholder="Name" />
                <textarea placeholder="Description"></textarea>
                <button type="submit">Submit</button>
              </form>
            </div>
          </AccordionDetails>
        </Accordion>
      );
      
      expect(screen.getByPlaceholderText('Name')).toBeInTheDocument();
      expect(screen.getByPlaceholderText('Description')).toBeInTheDocument();
      expect(screen.getByRole('button', { name: /submit/i })).toBeInTheDocument();
    });
  });

  // Snapshots
  describe('Snapshots', () => {
    it('matches snapshot for default props', () => {
      const { container } = renderWithTheme(
        <Accordion>
          <AccordionSummary>Default Summary</AccordionSummary>
          <AccordionDetails>Default Details</AccordionDetails>
        </Accordion>
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches snapshot for all variants', () => {
      const variants: AccordionProps['variant'][] = [
        'standard', 'outlined', 'filled', 'elevated', 'ghost'
      ];

      variants.forEach((variant) => {
        const { container } = renderWithTheme(
          <Accordion variant={variant}>
            <AccordionSummary>{variant} Summary</AccordionSummary>
            <AccordionDetails>{variant} Details</AccordionDetails>
          </Accordion>
        );
        expect(container.firstChild).toMatchSnapshot(`variant-${variant}`);
      });
    });

    it('matches snapshot for expanded state', () => {
      const { container } = renderWithTheme(
        <Accordion expanded>
          <AccordionSummary>Expanded Summary</AccordionSummary>
          <AccordionDetails>Expanded Details</AccordionDetails>
        </Accordion>
      );
      expect(container.firstChild).toMatchSnapshot('expanded-state');
    });
  });
});