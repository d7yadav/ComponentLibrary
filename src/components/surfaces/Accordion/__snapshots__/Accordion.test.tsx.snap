// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Accordion Component > Snapshots > matches snapshot for all variants > variant-elevated 1`] = `
<style>
  
  @keyframes accordion-loading-spin {
    from {
      transform: rotate(0deg);
    }
    to {
      transform: rotate(360deg);
    }
  }

</style>
`;

exports[`Accordion Component > Snapshots > matches snapshot for all variants > variant-filled 1`] = `
<style>
  
  @keyframes accordion-loading-spin {
    from {
      transform: rotate(0deg);
    }
    to {
      transform: rotate(360deg);
    }
  }

</style>
`;

exports[`Accordion Component > Snapshots > matches snapshot for all variants > variant-ghost 1`] = `
<style>
  
  @keyframes accordion-loading-spin {
    from {
      transform: rotate(0deg);
    }
    to {
      transform: rotate(360deg);
    }
  }

</style>
`;

exports[`Accordion Component > Snapshots > matches snapshot for all variants > variant-outlined 1`] = `
<style>
  
  @keyframes accordion-loading-spin {
    from {
      transform: rotate(0deg);
    }
    to {
      transform: rotate(360deg);
    }
  }

</style>
`;

exports[`Accordion Component > Snapshots > matches snapshot for all variants > variant-standard 1`] = `
<style>
  
  @keyframes accordion-loading-spin {
    from {
      transform: rotate(0deg);
    }
    to {
      transform: rotate(360deg);
    }
  }

</style>
`;

exports[`Accordion Component > Snapshots > matches snapshot for default props 1`] = `
<style>
  
  @keyframes accordion-loading-spin {
    from {
      transform: rotate(0deg);
    }
    to {
      transform: rotate(360deg);
    }
  }

</style>
`;

exports[`Accordion Component > Snapshots > matches snapshot for expanded state > expanded-state 1`] = `
<style>
  
  @keyframes accordion-loading-spin {
    from {
      transform: rotate(0deg);
    }
    to {
      transform: rotate(360deg);
    }
  }

</style>
`;
